dot/network/discovery.go

```diff
@@ -208,16 +210,12 @@ func (d *discovery) findPeers(ctx context.Context) {
 
 			logger.Trace("found new peer via DHT", "peer", peer.ID)
 
-			// found a peer, try to connect if we need more peers
-			if len(d.h.Network().Peers()) < d.maxPeers {
-				err = d.h.Connect(d.ctx, peer)
-				if err != nil {
-					logger.Trace("failed to connect to discovered peer", "peer", peer.ID, "err", err)
-				}
-			} else {
-				d.h.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
-				return
-			}
+			d.h.Peerstore().AddAddrs(peer.ID, peer.Addrs, peerstore.PermanentAddrTTL)
+			d.handler.AddPeer(0, peer.ID)
 		}
 	}
 }
+
+func (d *discovery) findPeer(peerID peer.ID) (peer.AddrInfo, error) {
+	return d.dht.FindPeer(d.ctx, peerID)
```

Comment:
Well it depends, sometimes both are needed.

For example since you can't fiddle with Go's standard library HTTP goroutine dispatcher (and pass a context to each goroutine starting), you might want to cancel the context for all requests being processed when the entire server should shutdown, and so you need to set a context as field of your server/handler receiver struct.